## 🕔 Block 5 — File Handling with NumPy + Data Simulation (1.5 hours)

### Learn

* Load/save `.csv`, `.txt`, `.npy`
* Random distributions

### Practice Tasks

31. Save array to CSV
32. Load CSV into NumPy
33. Generate normal distribution
34. Generate binomial distribution
35. Simulate coin toss experiment
36. Simulate dice roll experiment


In [1]:
import numpy as np

# Step 1: Create array
arr = np.array([
    [1.0, 2.0, 3.0],
    [4.0, 5.0, 6.0]
])

# Step 2: Save to CSV
np.savetxt("data.csv", arr, delimiter=",", fmt="%.2f")


In [2]:
loaded = np.loadtxt("data.csv", delimiter=",")

print(loaded)
print(type(loaded))


[[1. 2. 3.]
 [4. 5. 6.]]
<class 'numpy.ndarray'>


In [3]:
np.savetxt("data.txt", arr)
loaded = np.loadtxt("data.txt")


In [4]:
np.save("data.npy", arr)


In [5]:
import numpy as np

np.random.seed(42)

samples = np.random.normal(
    loc=0.0,      # mean
    scale=1.0,    # std deviation
    size=10
)

print(samples)


[ 0.49671415 -0.1382643   0.64768854  1.52302986 -0.23415337 -0.23413696
  1.57921282  0.76743473 -0.46947439  0.54256004]


In [6]:
samples = np.random.binomial(
    n=10,     # number of trials
    p=0.5,     # probability of success
    size=100
)

print(samples)


[4 4 4 5 5 4 5 3 4 4 5 6 4 5 5 2 5 3 3 8 8 6 4 3 6 5 3 5 2 7 4 6 4 5 5 4 8
 6 7 7 5 7 3 4 2 4 5 4 7 4 4 5 3 6 3 8 6 4 1 6 6 6 6 3 4 3 7 6 4 3 4 4 6 6
 7 5 3 6 6 5 6 5 5 5 2 3 2 6 4 5 7 4 5 6 4 3 4 3 7 6]


In [ ]:
tosses = np.random.binomial(
    n=1,
    p=0.5,
    size=1000
)

print("Number of heads:", np.sum(tosses))


Number of heads: 50116


In [28]:
rolls = np.random.randint(
    1,   # low
    7,   # high (exclusive)
    size=1000
)

print("Mean roll:", np.mean(rolls))


Mean roll: 3.455


In [29]:
dice1 = np.random.randint(1, 7, size=1000)
dice2 = np.random.randint(1, 7, size=1000)

sums = dice1 + dice2
print(sums)

print("Average sum:", np.mean(sums))


[ 7 12  8  9  3  7  8  8  9  6  7 11  7  3  6 10  4  8  6  6  9 12  7  5
  8 11 11  8  6  2  6  5  5  6  8  5  6  8 11  7  4  8  5  4  9  6 11 10
  8  5  7  3  8  7  5  5 10  5  7  9  7  8 11  8  3  6  7  5 12  9  7  5
  5  4  6  6  9  3  7  6  6  7  5  4  7  7  7  7  7  4  6  7  5  8  7  5
  3  6  7  9  8  4  6  5 12  4  7 12  7 10  9  6 10  9  5 11  3  4  5  9
  9  6  4 10  8  6  8  8  5  7  7 12  9  7  9  5 10  5  6  7 10  3  9 10
  5  8  5  8  7 11  7  7  4  5  4  8 10  8  9  7  5  8  5 11  7 11  8  6
 10  4  6  7  6  9  5 11 10  6  2  6 12  7  8  6  8  9  9  4  3  9  7  8
 11  8  4  7 11 10  9  8  7  9  8  3  9  9  5  7  4  6  5  9  7  8  4  3
  4  9  9  6  9 10  9  5  7 10  5  8  7  6  2  5 10  6  8  8  6  6  5  5
  8  8  6 10  7  8  7  3  5  7  6  7  9  4  6  6 10  6  6  9  8  8 11 11
  7 10  8  9  5  5  4  7  8  6  8  8  9  4 11  8  6  6  6  9  9  5  4  3
  4  5 10  4  8  6  9  6  9  9  5  8  7 12  7  4  5  9  5  7  7  6  7  5
  6  5  3  7  9  5  4  6  5  8 12 12  5 11  7  4  6

# 🕔 PART 1 — File Handling with NumPy

---

# 🔹 1️⃣ Saving & Loading `.csv`

CSV = plain text file with comma-separated values.

---

## ✅ 31. Save Array to CSV

### 🧠 Concept

We convert NumPy array → text format.

### 🧱 Structure

```python
import numpy as np

# Step 1: Create array
arr = np.array([
    [__, __, __],
    [__, __, __]
])

# Step 2: Save to CSV
np.savetxt("data.csv", arr, delimiter=",", fmt="%.2f")
```

### 🔎 Important Parameters

* `delimiter=","` → CSV format
* `fmt="%.2f"` → 2 decimal formatting

---

## ✅ 32. Load CSV into NumPy

### 🧱 Structure

```python
loaded = np.loadtxt("data.csv", delimiter=",")

print(loaded)
print(type(loaded))
```

---

# 🔹 2️⃣ Save/Load `.txt`

Same as CSV but different delimiter.

```python
np.savetxt("data.txt", arr)
loaded = np.loadtxt("data.txt")
```

---

# 🔹 3️⃣ Save/Load `.npy` (Binary Format)

⚡ Faster and preserves dtype exactly.

### Save

```python
np.save("data.npy", arr)
```

### Load

```python
loaded = np.load("data.npy")
```

Use `.npy` when working with ML datasets.

---

# 🕔 PART 2 — Random Distributions (Data Simulation)

Now we move into probability + simulation.

---

# 🔹 33. Generate Normal Distribution

### 🧠 Mathematics

Normal distribution:

[
X \sim N(\mu, \sigma^2)
]

Where:

* ( \mu ) = mean
* ( \sigma ) = standard deviation

### 🧱 Structure

```python
import numpy as np

np.random.seed(42)

samples = np.random.normal(
    loc=__,      # mean
    scale=__,    # std deviation
    size=__
)

print(samples)
```

---

### 📊 Interpretation

* Symmetric bell curve
* Used in ML, noise modeling, statistics

---

# 🔹 34. Generate Binomial Distribution

### 🧠 Mathematics

[
X \sim Bin(n, p)
]

* n = number of trials
* p = probability of success

Counts number of successes.

---

### 🧱 Structure

```python
samples = np.random.binomial(
    n=__,     # number of trials
    p=__,     # probability of success
    size=__
)

print(samples)
```

---

# 🔹 35. Simulate Coin Toss Experiment

Coin toss = Binomial with:

* n = 1
* p = 0.5

---

## 🎯 Single Toss Many Times

```python
tosses = np.random.binomial(
    n=1,
    p=0.5,
    size=1000
)

print("Number of heads:", np.sum(tosses))
```

---

## 🎯 Simulate 10 Tosses Per Experiment

```python
experiments = np.random.binomial(
    n=10,
    p=0.5,
    size=1000
)

print("Average heads:", np.mean(experiments))
```

---

### 🧠 Why Binomial?

Each toss:

* Independent
* Two outcomes
* Fixed probability

That’s exactly binomial model.

---

# 🔹 36. Simulate Dice Roll Experiment

Dice = Uniform discrete distribution.

---

## 🎯 Roll Dice 1000 Times

```python
rolls = np.random.randint(
    1,   # low
    7,   # high (exclusive)
    size=1000
)

print("Mean roll:", np.mean(rolls))
```

Expected mean ≈ 3.5

---

## 🎯 Simulate 2 Dice Sum

```python
dice1 = np.random.randint(1, 7, size=1000)
dice2 = np.random.randint(1, 7, size=1000)

sums = dice1 + dice2

print("Average sum:", np.mean(sums))
```

Expected ≈ 7

---

# 🔥 Deep Understanding

### Why Simulation Matters

Instead of solving probability formulas manually:

We approximate using large random samples.

Law of Large Numbers:

[
\bar{X} \rightarrow E[X]
]

As sample size increases.

---

# 🧠 Real-World Use Cases

| Distribution    | Used In                      |
| --------------- | ---------------------------- |
| Normal          | Measurement noise, ML errors |
| Binomial        | A/B testing, classification  |
| Uniform         | Random sampling              |
| Dice simulation | Monte Carlo methods          |

---

# 🚀 Practice Strategy (Important)

For each task:

1. Generate 10 samples
2. Generate 10,000 samples
3. Compare mean with theoretical value
4. Plot histogram (optional but powerful)

---

# 🏆 Mini Challenge

Try this:

* Simulate 10,000 coin tosses
* Estimate probability of getting >7 heads in 10 tosses
* Compare with theoretical binomial formula

---

If you want next, I can:

* 📊 Teach Monte Carlo simulation
* 🧠 Connect distributions to ML loss functions
* 🔥 Create a mini data simulation project
* 📈 Teach histogram & visualization

Tell me your next move.
